src/compiler/checker.ts

```diff
@@ -9514,100 +9674,140 @@ module ts {
-
-                let iteratorFunction = getTypeOfPropertyOfType(iterable, getPropertyNameForKnownSymbolName("iterator"));
-                if (iteratorFunction && allConstituentTypesHaveKind(iteratorFunction, TypeFlags.Any)) {
-                    return undefined;
+                    typeAsIterable.iterableElementType = (<GenericType>iterable).typeArguments[0];
                 }
+                else {
+                    let iteratorFunction = getTypeOfPropertyOfType(iterable, getPropertyNameForKnownSymbolName("iterator"));
+                    if (iteratorFunction && allConstituentTypesHaveKind(iteratorFunction, TypeFlags.Any)) {
+                        return undefined;
+                    }
 
-                let iteratorFunctionSignatures = iteratorFunction ? getSignaturesOfType(iteratorFunction, SignatureKind.Call) : emptyArray;
-                if (iteratorFunctionSignatures.length === 0) {
-                    if (errorNode) {
-                        error(errorNode, Diagnostics.Type_must_have_a_Symbol_iterator_method_that_returns_an_iterator);
+                    let iteratorFunctionSignatures = iteratorFunction ? getSignaturesOfType(iteratorFunction, SignatureKind.Call) : emptyArray;
+                    if (iteratorFunctionSignatures.length === 0) {
+                        if (errorNode) {
+                            error(errorNode, Diagnostics.Type_must_have_a_Symbol_iterator_method_that_returns_an_iterator);
+                        }
+                        return undefined;
                     }
-                    return undefined;
-                }
 
-                let iterator = getUnionType(map(iteratorFunctionSignatures, getReturnTypeOfSignature));
-                if (allConstituentTypesHaveKind(iterator, TypeFlags.Any)) {
-                    return undefined;
+                    typeAsIterable.iterableElementType = getElementTypeFromIterator(getUnionType(map(iteratorFunctionSignatures, getReturnTypeOfSignature)), errorNode);
                 }
+            }
 
-                let iteratorNextFunction = getTypeOfPropertyOfType(iterator, "next");
-                if (iteratorNextFunction && allConstituentTypesHaveKind(iteratorNextFunction, TypeFlags.Any)) {
-                    return undefined;
+            return typeAsIterable.iterableElementType;
+        }
+
+        function getElementTypeFromIterator(iterator: Type, errorNode: Node): Type {
+            // This function has very similar logic as getElementTypeFromIterable, except that it operates on
+            // Iterators instead of Iterables. Here is the structure:
+            //
+            //  { // iterator
+            //      next: { // iteratorNextFunction
+            //          (): { // iteratorNextResult
+            //              value: T // iteratorNextValue
+            //          }
+            //      }
+            //  }
+            //
+            if (allConstituentTypesHaveKind(iterator, TypeFlags.Any)) {

```

Comment:
I find this line odd; we've discussed offline, we need to see if we can have an un-reduced union with `any`